### PR TITLE
Add header support for BatchedProducer

### PR DESCRIPTION
**Pull request by deviousasti**

### CODE REVIEW

**Comment by bartelink** [comment]
I'm not clear what the actual problem being solved is here (can you describe it in the overview please) ?

File: src/FsKafka/FsKafka.fs [hunk 1]
```diff
@@ -95,35 +95,56 @@ type KafkaProducerConfig private (inner, bootstrapServers : string) =
         customize |> Option.iter (fun f -> f c)
         KafkaProducerConfig(c, bootstrapServers)
 
+/// A message header consists of a string key and a byte-encoded value
+type MessageHeaders = seq<string * byte[]>
+
+module private Message =
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:101]: while this is arguably clearer, it allocates a Headers object in the `noHeaders` case, and adds lots of code - is there something being achieved to counterbalance this ?

File: src/FsKafka/FsKafka.fs [hunk 2]
```diff
@@ -168,11 +194,20 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
+
+        for message in kvhBatch do
+            inner.Produce(topic, Message.create message, deliveryHandler=handler)
+
         inner.Flush(ct)
         return! Async.AwaitTaskCorrect tcs.Task }
 
+    /// <summary>
+    /// Produces a batch of supplied key/value messages. 
+    /// See the other overload.
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:206]: can you express this more directy?

File: src/FsKafka/FsKafka.fs [hunk 3]
```diff
@@ -168,11 +194,20 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
+    /// </summary>
+    member __.ProduceBatch(messageBatch : seq<string * string>) : Async<DeliveryReport<string,string>[]> = 
+        __.ProduceBatch(messageBatch |> Seq.map(fun (key, value) -> key, value, Message.noHeaders))
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:209]: Seq.map this to Message

File: src/FsKafka/FsKafka.fs [hunk 4]
```diff
@@ -141,20 +162,25 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
     ///    1. if there is an immediate local config issue
     ///    2. upon receipt of the first failed `DeliveryReport` (NB without waiting for any further reports, which can potentially leave some results in doubt should a 'batch' get split) </throws>
     /// <remarks>
     ///    Note that the delivery and/or write order may vary from the supplied order unless `maxInFlight` is 1 (which massively constrains throughput).
-    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.<remarks/>
-    member __.ProduceBatch(keyValueBatch : (string * string)[]) : Async<DeliveryReport<string,string>[]> = async {
-        if Array.isEmpty keyValueBatch then return [||] else
+    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.
+    /// </remarks>
+    member __.ProduceBatch(messageBatch : seq<string * string * MessageHeaders>) : Async<DeliveryReport<string,string>[]> = async {
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:175]: make this take a `Message` seq and have an overload map from a seq of triples

File: src/FsKafka/FsKafka.fs [hunk 5]
```diff
@@ -95,35 +95,56 @@ type KafkaProducerConfig private (inner, bootstrapServers : string) =
...
-        | Some hs ->
-            let messageHeaders = Headers()
-            for k, v in hs do messageHeaders.Add(k, v)
-            message.Headers <- messageHeaders
+    /// There's no assurance of ordering [without dropping `maxInFlight` down to `1` and annihilating throughput].
+    /// Thus its critical to ensure you don't submit another message for the same key until you've had a success / failure 
+    /// response from the call.
+    /// </remarks>
+    member __.ProduceAsync(key : string, value: string, headers : MessageHeaders) : Async<DeliveryResult<string, string>> =
+        let message = Message.create(key, value, headers)
```
[the remark this answers is not in the document]
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:136]: On reflection, if you provide an overload that Produces a `Message` then you can split the headers/no-headers handling into two overloads that delegate to that with specific code constructing a Message directly

**Comment by bartelink** [comment]
> I'm not clear what the actual problem being solved is here (can you describe it in the overview please) ?
Hm, reading the full PR makes it make more sense

Reading it top to bottom, it seems there are lots of egregious and subjective breaking changes to the single item producer overloads.

I now see that this is a result of generalizing the batch approach

The crux of my comments overall is that the better approach here, is to provide overloads that take a `Message`, and inline the creation of those in the relevant into more lightweight methods that compose those, rather than forcing the higher levels to pass in Null Objects

File: src/FsKafka/FsKafka.fs [hunk 6]
```diff
@@ -95,35 +95,56 @@ type KafkaProducerConfig private (inner, bootstrapServers : string) =
         customize |> Option.iter (fun f -> f c)
         KafkaProducerConfig(c, bootstrapServers)
 
+/// A message header consists of a string key and a byte-encoded value
+type MessageHeaders = seq<string * byte[]>
```
[the remark this answers is not in the document]
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Removed

File: src/FsKafka/FsKafka.fs (hunk 5 above)
[the remark this answers is not in the document]
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Did it that way

File: src/FsKafka/FsKafka.fs (hunk 3 above)
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Done

**Comment by deviousasti** [comment]
@bartelink Please review. If we can merge this in, I can start on removing Kafka0.

File: src/FsKafka/FsKafka.fs [hunk 7]
```diff
@@ -141,20 +163,25 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
     ///    2. upon receipt of the first failed `DeliveryReport` (NB without waiting for any further reports, which can potentially leave some results in doubt should a 'batch' get split) </throws>
     /// <remarks>
     ///    Note that the delivery and/or write order may vary from the supplied order unless `maxInFlight` is 1 (which massively constrains throughput).
-    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.<remarks/>
-    member __.ProduceBatch(keyValueBatch : (string * string)[]) : Async<DeliveryReport<string,string>[]> = async {
-        if Array.isEmpty keyValueBatch then return [||] else
+    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.
+    /// </remarks>
+    member __.ProduceBatch(messageBatch : seq<Message<_, _>>) : Async<DeliveryReport<string,string>[]> = async {
+        match Array.ofSeq messageBatch with
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:177]: if you are going to materialize it immediately like this, better to require that instead

File: src/FsKafka/FsKafka.fs [hunk 8]
```diff
@@ -168,11 +195,27 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
+
         inner.Flush(ct)
         return! Async.AwaitTaskCorrect tcs.Task }
 
+    /// <summary>
+    /// Produces a batch of supplied key/value messages. 
+    /// See the other overload.
+    /// </summary>
+    member __.ProduceBatch(messageBatch : seq<string * string>) : Async<DeliveryReport<string,string>[]> = 
+        __.ProduceBatch(messageBatch |> Seq.map Message.create)
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:210]: Array.map for same reasons

File: src/FsKafka/FsKafka.fs [hunk 9]
```diff
@@ -95,35 +95,57 @@ type KafkaProducerConfig private (inner, bootstrapServers : string) =
         customize |> Option.iter (fun f -> f c)
         KafkaProducerConfig(c, bootstrapServers)
 
+
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:98]: remove blank line (nits are fine creatures!)

File: src/FsKafka/FsKafka.fs [hunk 10]
```diff
@@ -95,35 +95,57 @@ type KafkaProducerConfig private (inner, bootstrapServers : string) =
...
+        Message<_,_>(Key=key, Value=value)
+
+    let createWithHeaders(key: string, value: string, headers : seq<string * byte[]>) =
+        let message = Message<_,_>(Key=key, Value=value, Headers = Headers())
+        for header in headers do
+            message.Headers.Add(Header header)
+
+        message
+
+    let noHeaders : seq<string * byte[]> = Seq.empty
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:111]: dead. also remove L108!

File: src/FsKafka/FsKafka.fs (hunk 1 above)
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:99]: Nice refactor!

**Comment by bartelink** [comment]
Looks great now, thanks for finding a good solution @deviousasti

File: src/FsKafka/FsKafka.fs (hunk 8 above)
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:210]: On reflection, taking a seq here would make sense, but the Array.ofSeq should be in here

**Comment by bartelink** [comment]
Oh yeah, needs changelog entry too

File: src/FsKafka/FsKafka.fs (hunk 7 above)
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:175]: no array.ofSeq now?

File: CHANGELOG.md [hunk 11]
```diff
@@ -13,11 +13,18 @@ The `Unreleased` section name is replaced by the expected version of next releas
...
+- Added overload of `BatchedProducer` to support passing in `Message[]`
+
 ### Changed
 ### Removed
 ### Fixed
 
+- xmldocs for `BatchedProducer`
+
 <a name="1.6.2"></a>
+
```
Comment (bartelink) [inline CHANGELOG.md:27]: remove

File: CHANGELOG.md [hunk 12]
```diff
@@ -13,11 +13,18 @@ The `Unreleased` section name is replaced by the expected version of next releas
...
 ### Added
+
+- Added overload of `BatchedProducer` to support message headers
+- Added overload of `BatchedProducer` to support passing in `Message[]`
+
 ### Changed
 ### Removed
 ### Fixed
 
+- xmldocs for `BatchedProducer`
```
Comment (bartelink) [inline CHANGELOG.md:24]: I tend to leave this out - though if the F12 and sourcelink starts working as a result of this, it's almost a new _Added_ class thing!

File: CHANGELOG.md [hunk 13]
```diff
@@ -138,7 +145,6 @@ The `Unreleased` section name is replaced by the expected version of next releas
 ### Added
 
 - include `partition` property as `LogContext.PushProperty` when dispatching a handler invocation [#60](https://github.com/jet/FsKafka/pull/60)
-
```
Comment (bartelink) [inline CHANGELOG.md:141]: I tend to leave spaces here before header as it sometimes does not render right?

File: src/FsKafka/FsKafka.fs (hunk 10 above)
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:109]: isnt this dead?

File: CHANGELOG.md (hunk 12 above)
Comment (deviousasti) [reply CHANGELOG.md:24]: The thing was just missing the summary tag!

File: src/FsKafka/FsKafka.fs (hunk 7 above)
Comment (bartelink) [reply src/FsKafka/FsKafka.fs:175]: now you can rever it back to L152 - array.isEmpty check

File: src/FsKafka/FsKafka.fs [hunk 14]
```diff
@@ -168,11 +193,27 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
+    /// </summary>
+    member __.ProduceBatch(messageBatch : seq<string * string>) : Async<DeliveryReport<string,string>[]> = 
+        __.ProduceBatch(messageBatch |> Seq.map Message.create |> Array.ofSeq)
+
+    /// <summary>
+    /// Produces a batch of messages with supplied key/value/headers. 
+    /// See the other overload.
+    /// </summary>
+    member __.ProduceBatch(messageBatch : seq<string * string * seq<string * byte[]>>) : Async<DeliveryReport<string,string>[]> = 
+        __.ProduceBatch(messageBatch |> Seq.map Message.createWithHeaders |> Array.ofSeq)
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:215]: maybe an array comprehension might look better and be same perf?

File: src/FsKafka/FsKafka.fs (hunk 7 above)
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Right.

File: src/FsKafka/FsKafka.fs [hunk 15]
```diff
@@ -141,20 +161,22 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
     /// <remarks>
     ///    Note that the delivery and/or write order may vary from the supplied order unless `maxInFlight` is 1 (which massively constrains throughput).
-    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.<remarks/>
-    member __.ProduceBatch(keyValueBatch : (string * string)[]) : Async<DeliveryReport<string,string>[]> = async {
-        if Array.isEmpty keyValueBatch then return [||] else
+    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.
+    /// </remarks>
+    member __.ProduceBatch(messageBatch : Message<_, _>[]) : Async<DeliveryReport<string,string>[]> = async {
 
         let! ct = Async.CancellationToken
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:177]: need to do early exit if batch empty ? people could be leaning on that semantic even if it did work (yes, arguably there should be a test for it)

File: src/FsKafka/FsKafka.fs [hunk 16]
```diff
@@ -141,14 +161,17 @@ type BatchedProducer private (inner : IProducer<string, string>, topic : string)
...
     /// <throws>
     ///    1. if there is an immediate local config issue
     ///    2. upon receipt of the first failed `DeliveryReport` (NB without waiting for any further reports, which can potentially leave some results in doubt should a 'batch' get split) </throws>
     /// <remarks>
     ///    Note that the delivery and/or write order may vary from the supplied order unless `maxInFlight` is 1 (which massively constrains throughput).
-    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.<remarks/>
-    member __.ProduceBatch(keyValueBatch : (string * string)[]) : Async<DeliveryReport<string,string>[]> = async {
+    ///    Thus it's important to note that supplying >1 item into the queue bearing the same key without maxInFlight=1 risks them being written out of order onto the topic.
+    /// </remarks>
+    member __.ProduceBatch(keyValueBatch : Message<_, _>[]) : Async<DeliveryReport<string,string>[]> = async {
```
Comment (bartelink) [inline src/FsKafka/FsKafka.fs:174]: wait, the rename was nice and OK, (and not crazy breaking for my taste) !

Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Went all in on diff minimization

File: src/FsKafka/FsKafka.fs (hunk 15 above)
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs:177]: Will add that in after we kill Kafka0

File: src/FsKafka/FsKafka.fs (hunk 10 above)
Comment (deviousasti) [reply src/FsKafka/FsKafka.fs]: Killed it

**Comment by bartelink** [comment]
Thanks, this looks great, and wonderful to see actual meaningful tests

**Pull request merged**